### DISCUSSION TITLE
Taskbar window caption shows [WARN: COPY MODE]

### DISCUSSION THREAD

**Discussion by badjames** [microsoft/wslg#312] [Q&A]
This appears on all WSLg window caption in the taskbar, but not on the actual window titles.

I have heard mentioned that this could be because WSLg is using RAIL rather than VAIL for the RDP transport.

I also noticed the following in the weston.log file which perhaps confirms that:

[10:40:49.922] WSL2_SHARED_MEMORY_MOUNT_POINT is not set.

In addition, "/mnt/shared_memory" is not present.

I am also getting quite poor 3D framerates despite having HW acceleration working (glxinfo shows renderer as "D3D12 (NVIDIA GTX 1070).  Might this be due to the use of RAIL sending bits over the wire, rather than using shared memory as in VAIL?  If so, how do I get VAIL working?

Update 18/06: I updated WSL (wsl --update) and this removed the "[WARN: COPY MODE]".  However, OpenGL performance hasn't improved and I still don't see a /mnt/shared_memory.

**Comment by onomatopellan** [comment]
You need to upgrade to mesa 21.x
```
sudo add-apt-repository ppa:kisak/kisak-mesa
sudo apt-get update && sudo apt upgrade
```

**Reply by badjames** [reply]
Thanks, see the "update" at the bottom of my original post.

**Reply by onomatopellan** [reply]
You can see the `/mnt/shared_memory` inside the system distro. 
`wsl.exe --system -d $WSL_DISTRO_NAME`
`ls -la /mnt/`

The reason OpenGL performance is not good (but way better than sotfware rendering) is explained here https://devblogs.microsoft.com/commandline/wslg-architecture/#hardware-accelerated-opengl

**Reply by badjames** [reply]
That is very useful to know, thanks!

I do have vGPU and the latest Mesa, but perhaps it is the system memory transfer that is killing the frames.

**Reply by Pierre-Datasolution** [reply]
Worked for me. Thanks

**Reply by palapapa** [reply]
For those who thought that this didn't work, remember to run `wsl --shutdown` and then restart.

**Reply by biggestsonicfan** [reply]
Doesn't quite help for those not on a debian distro and who already have Mesa 23 or higher installed.

**Reply by RogerJL** [reply]
Works on Microsoft standard WSL distro
this fixes garbled qt graphics output (probably get window size wrong - clicking on widow helped but interactivity was lost)
Now this works, thanks!

**Reply by crystalfp** [reply]
Unfortunately not working.
![image](https://github.com/microsoft/wslg/assets/865866/cec1c88c-217a-4d65-a857-3fa05f62da95)
And on console:
```
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: Some incorrect rendering might occur because the selected Vulkan device (Microsoft Direct3D12 (NVIDIA Quadro T1000)) doesn't support base Zink requirements: feats.features.logicOp have_EXT_custom_border_color have_EXT_line_rasterization
glx: failed to create drisw screen
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: dzn is not a conformant Vulkan implementation, testing use only.
WARNING: Some incorrect rendering might occur because the selected Vulkan device (Microsoft Direct3D12 (NVIDIA Quadro T1000)) doesn't support base Zink requirements: feats.features.logicOp have_EXT_custom_border_color have_EXT_line_rasterization
glx: failed to create drisw screen
```
I run Ubuntu 22.04

**Reply by antonio-petricca** [reply]
Reappared after rebooting.

**Comment by spronovo** [comment]
This error message does indeed mean RDP is running in RAIL mode (copy pixels from Linux to Windows over the RDP transport) instead of VAIL mode (share memory between Linux and Windows).

This is actually independent of mesa or 3D acceleration / vGPU drivers. The WSLg shared memory uses virtio-fs which has been enabled in the WSL 2 Linux kernel from version 5.10.16+.

Could you verify the version of the Linux kernel you are running:

```
spronovo@OFFICE:~$ uname -r
5.10.16.3-microsoft-standard-WSL2
```

If you are running a version older than 5.10.16, shut down WSL ```wsl --shutdown``` and update ```wsl --update```. If that doesn't pick up any update, double check that your .wslconfig isn't setup to use a private kernel (```kernel=xyz```) as this would load your private and skip any update we send down.

If you are on 5.10.16+ and you are still seeing that error, we would need to get the dmseg log from the Linux kernel boot session to see what failed when attempting to initialize virtio-fs. To gather this, you can enable the WSL debug console. In your ```c:\users\__user_name__\.wslconfig``` enter the following content:

```
[wsl2]
debugConsole = true
```

Restart wsl ```wsl --shutdown```, and on restart you should see a console showing up each time you launch WSL and we boot the Linux kernel. Share the output you get there.

**Reply by badjames** [reply]
Thanks for the detailed response!

I was indeed able to resolve the issue via wsl --update

**Reply by garyo** [reply]
I'm having the same issue. WSL2, Ubuntu 20.04, `uname -r` reports `5.10.60.1-microsoft-standard-WSL2`.
I recorded my startup debug log in this gist: https://gist.github.com/garyo/f386d9ee36729e58b72b098b2b07d51a
I do not have a `/mnt/shared_memory` in my distro.
My host CPU is AMD x86_64, Ryzen Threadripper 1950X 16-core.

Weirdly, it doesn't always happen. I see it on some apps always (emacs), some apps only sometimes.

**Comment by fredboulanger** [comment]
So at my end running 5.10.43.3-microsoft-standard-WSL2 on ARM surfacepro x. I get WARN:COPY caption on some apps, and for example vscode (installed in wsl2) doesn't. The issue with the apps showing warn:copy is they don't respond to window's window management  (for example they don't snap) and they seem much slower.

Has anyone ran into this issue, and found a fix?

**Reply by spronovo** [reply]
Hi @fredboulanger, this is because on ARM platform our shared memory optimization aren't currently supported so WSLg runs in RAIL mode (copy pixels over the RDP channel). We have an updated WSL Linux Kernel for ARM in the pipeline that will be adding the missing support, after which this warning will be going away and your system will be running the more performant VAIL mode.

In term of window snapping, that's a known limitation at the moment as window movement and resize are handled on the Linux side of the world. This is something we would like to improve in a future.

**Comment by fredboulanger** [comment]
Thanks for the thorough response.

Re window snapping. What I fail to understand if the lack of consistency between Linux applications. For context Firefox doesn't behave well but Spyder from Anaconda behaves like a Windows application. Maybe this is what you mean when you say it's a bug, the inconsistency between apps.


Thanks again
Fred

Frederic Boulanger / 613.866 6997
________________________________

**Comment by mitsos1os** [comment]
This was still valid for me and the [solution](https://github.com/microsoft/wslg/discussions/312#discussioncomment-913719) proposed fixes the problem.

OS: Win 11 22H2
WSL2: Ubuntu 20.04.5
kernel: 5.15.79.1-microsoft-standard-WSL2

**Comment by ericblade** [comment]
I have been working on a custom distribution that i'm trying to get working in WSL, for the last several days, and everything was running fine -- and then suddenly it wasn't.  WSL, as well as some windows processes, slowed to a crawl, and my 1 running WSLg process is showing [WARN: COPY MODE] on it.  a wsl --shutdown hasn't cleared it, going to try rebooting windows.

Definitely is running mesa 24, using llvmpipe as the d3d12 drivers don't seem to like Apple's AMD cards.

update: rebooting fixed it.  Suspecting some sort of resource leak inside my wsl distro, i attempted relaunching every single thing inside it again, which resulted in pulseaudio breaking instead of wslg, necessitating a wsl --shutdown to get back to anything working

**Comment by yousecjoe** [comment]
I never seen this issue in all the years using WSLg, until today. 
`6.18.33.2-microsoft-standard-WSL2`
`VERSION="24.04.4 LTS (Noble Numbat)"`

**Comment by Kamicosi** [comment]
Fix temporarily with
```
sudo mkdir -p /mnt/shared_memory
sudo mount -t tmpfs tmpfs /mnt/shared_memory
```

**Comment by ambs** [comment]
Nothing in this thread works. The only approach is shutting down and restarting wsl, which is a shitty solution when you have other terminals doing relevant stuff.

Running 6.18.33.2-microsoft-standard-WSL2, debian.

**Comment by robertbissonnette** [comment]
This [comment](https://github.com/microsoft/wslg/discussions/312#discussioncomment-913719) resolved a gitkraken wslg issue with WARN: COPY MODE and no ui window ever present. Only visible from the taskbar hover in windows.
